### ISSUE TITLE
Implement adaptive gradient clipping algorithm

### ISSUE THREAD

**Issue by lguerdan** [lguerdan/10-701-Project#3]
This task involves implementing the adaptive quantile clipping algorithm proposed by Andrew et al. ([https://arxiv.org/abs/1905.03871](https://arxiv.org/abs/1905.03871)). 

Additional information: 
- [Here](https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/dp_query/quantile_adaptive_clip_sum_query.py) is an existing implementation from Andrew et. al in Tensorflow. This implementation can be converted to PyTorch. The `get_noised_result(self, sample_state, global_state):` function in particular is a good resource to check. 
- Section 2 of the Andrew et. al. paper gives the equations for the cutoff update (also included in our mid-progress report). We could implement both the linear and geometric update variants. 
- [This loop](https://github.com/lguerdan/10-701-Project/blob/b3ff1c888b30e34c4526128599dea2222cf5db55/train.py#L126) in our `train.py` implementation is a good place to add this code because it is where we iterate over sample-wise gradients. 

To start with this task, it might be easiest to begin by storing the full gradient norm distribution and comparing the actual median to the median we estimate via their update equations. Once this looks correct, we could use the estimated median to update the norm clipping threshold.
